Should equal units always have equal hashes?

Consider composite units that have equal bases and powers and equal scale values, but different scale types:
```python
>>> from astropy import units as u
>>> from fractions import Fraction
>>> units = [u.CompositeUnit(scale, [u.m], [-1]) for scale in [2, 2.0, Fraction(2, 1)]]
>>> [type(unit.scale) for unit in units]
[<class 'int'>, <class 'float'>, <class 'fractions.Fraction'>]
```
All these units have identical string representations:
```python
>>> units
[Unit("2 / m"), Unit("2 / m"), Unit("2 / m")]
```
Furthermore, all the units are equal to each other:
```python
>>> units[0] == units[1]
True
>>> units[0] == units[2]
True
>>> units[1] == units[2]
True
```
But despite that, their hashes are not all equal:
```python
>>> [hash(unit) for unit in units]
[-7234605082436077977, 1904857767595994367, -7234605082436077977]
```
That can lead to very unexpected behavior:
```python
>>> from collections import Counter
>>> Counter(units)
Counter({Unit("2 / m"): 2, Unit("2 / m"): 1})
```
Should we allow such surprises? #16055 raised an analogous question about the powers of the components of a unit and there it was decided that if the powers of the components of a unit have equal values then they should also have the same types. This is enforced by https://github.com/astropy/astropy/blob/7368ab84abfeddf960b7315a12a565de948615de/astropy/units/utils.py#L263
Should something analogous be done with scales too?

> Simply remove the `str` of the scale passes all tests, but I worry that then also removing that of the powers, as in
> ```
> return hash((self.scale, *[x.name for x in self.bases], *self.powers))
> ```
> causes a test failure, since the hash is used inside _compose and apparently something goes wrong there.

There is indeed a failure:
```
    def test_compose_issue_579():
        unit = u.kg * u.s**2 / u.m
    
        result = unit.compose(units=[u.N, u.s, u.m])
    
        assert len(result) == 1
        assert result[0]._bases == [u.s, u.N, u.m]
>       assert result[0]._powers == [4, 1, -2]
E       assert [4, 1, -1] == [4, 1, -2]
```
I don't know if this is a CPython implementation detail or not, but at least in CPython the hash of an `int` is the `int` itself with one exception: the hash of `-1` is `-2` (it has something to do with some internal C code using `-1` as an error value, which prevents it from being used as a hash value). So if there is something that compares units only based on their hashes then powers of `-1` and `-2` become indistinguishable for it. But this would not be a concern for scales if we don't allow them to have negative values. Can magnitudes have negative scales?

--- a/astropy/units/core.py
+++ b/astropy/units/core.py
@@ -902,9 +902,7 @@ class UnitBase:
 
     @cached_property
     def _hash(self) -> int:
-        return hash(
-            (str(self.scale), *[x.name for x in self.bases], *map(str, self.powers))
-        )
+        return hash((self.scale, *[x.name for x in self.bases], *map(str, self.powers)))
 
     def __getstate__(self) -> dict[str, object]:
         # If we get pickled, we should *not* store the memoized members since

--- a/astropy/units/core.py
+++ b/astropy/units/core.py
@@ -1233,9 +1233,8 @@ class UnitBase:
             return all(base in namespace for base in unit.bases)
 
         unit = self.decompose()
-        key = hash(unit)
 
-        cached = cached_results.get(key)
+        cached = cached_results.get(unit)
         if cached is not None:
             if isinstance(cached, Exception):
                 raise cached
@@ -1244,7 +1243,7 @@ class UnitBase:
         # Prevent too many levels of recursion
         # And special case for dimensionless unit
         if depth >= max_depth:
-            cached_results[key] = [unit]
+            cached_results[unit] = [unit]
             return [unit]
 
         # Make a list including all of the equivalent units
@@ -1297,7 +1296,7 @@ class UnitBase:
         for final_result in final_results:
             if len(final_result):
                 results = final_results[0].union(final_results[1])
-                cached_results[key] = results
+                cached_results[unit] = results
                 return results
 
         partial_results.sort(key=operator.itemgetter(0))
@@ -1332,17 +1331,17 @@ class UnitBase:
                     subresults.add(factored)
 
             if len(subresults):
-                cached_results[key] = subresults
+                cached_results[unit] = subresults
                 return subresults
 
         if not is_final_result(self):
             result = UnitsError(
                 f"Cannot represent unit {self} in terms of the given units"
             )
-            cached_results[key] = result
+            cached_results[unit] = result
             raise result
 
-        cached_results[key] = [self]
+        cached_results[unit] = [self]
         return [self]
 
     def compose(

--- a/astropy/units/tests/test_units.py
+++ b/astropy/units/tests/test_units.py
@@ -1150,20 +1150,15 @@ def test_hash_represents_unit(unit, power):
     assert hash(tu2) == hash(unit)
 
 
-bug_revealing_xfail = pytest.mark.xfail(
-    reason="regression test to demonstrate an existing bug"
-)
-
-
 @pytest.mark.parametrize(
     "scale1, scale2",
     [
-        pytest.param(10, 10.0, marks=bug_revealing_xfail),
+        (10, 10.0),
         (2, Fraction(2, 1)),
-        pytest.param(4, 4 + 0j, marks=bug_revealing_xfail),
-        pytest.param(0.5, Fraction(1, 2), marks=bug_revealing_xfail),
+        (4, 4 + 0j),
+        (0.5, Fraction(1, 2)),
         (2.4, 2.4 + 0j),
-        pytest.param(Fraction(1, 4), 0.25 + 0j, marks=bug_revealing_xfail),
+        (Fraction(1, 4), 0.25 + 0j),
     ],
     ids=type,
 )

--- a/astropy/units/tests/test_units.py
+++ b/astropy/units/tests/test_units.py
@@ -1150,6 +1150,31 @@ def test_hash_represents_unit(unit, power):
     assert hash(tu2) == hash(unit)
 
 
+bug_revealing_xfail = pytest.mark.xfail(
+    reason="regression test to demonstrate an existing bug"
+)
+
+
+@pytest.mark.parametrize(
+    "scale1, scale2",
+    [
+        pytest.param(10, 10.0, marks=bug_revealing_xfail),
+        (2, Fraction(2, 1)),
+        pytest.param(4, 4 + 0j, marks=bug_revealing_xfail),
+        pytest.param(0.5, Fraction(1, 2), marks=bug_revealing_xfail),
+        (2.4, 2.4 + 0j),
+        pytest.param(Fraction(1, 4), 0.25 + 0j, marks=bug_revealing_xfail),
+    ],
+    ids=type,
+)
+def test_hash_scale_type(scale1, scale2):
+    # Regression test for #17820 - hash could depend on scale type, not just its value
+    unit1 = u.CompositeUnit(scale1, [u.m], [-1])
+    unit2 = u.CompositeUnit(scale2, [u.m], [-1])
+    assert unit1 == unit2
+    assert hash(unit1) == hash(unit2)
+
+
 @pytest.mark.skipif(not HAS_ARRAY_API_STRICT, reason="tests array_api_strict")
 def test_array_api_strict_arrays():
     # Ensure strict array api arrays can be passed in/out of Unit.to()

--- a/docs/changes/units/17822.bugfix.rst
+++ b/docs/changes/units/17822.bugfix.rst
@@ -1,0 +1,7 @@
+Previously the hashes of units could depend on the type of their scale factor,
+but now the hashes depend on the scale factor values, not the types.
+For example, previously a ``dict`` would consider two otherwise identical units
+with scale factors ``1`` (``int``) and ``1.0`` (``float``) as different keys,
+but now they are considered to be the same key.
+In practice this bug was rare because the number of units with a scale factor
+that is not a ``float`` is very small.

